clang/lib/CIR/Dialect/Transforms/RegionInlining.cpp

```diff
@@ -0,0 +1,308 @@
+    // TODO(CIR): stackSaveOp
+    
+    rewriter.create<mlir::cir::BrOp>(loc, mlir::ValueRange(), beforeBody);
+
+    // Replace the scopeop return with a branch that jumps out of the body.
+    // Stack restore before leaving the body region.
+    rewriter.setInsertionPointToEnd(afterBody);
+    if (auto yieldOp =
+            dyn_cast<mlir::cir::YieldOp>(afterBody->getTerminator())) {
+      rewriter.replaceOpWithNewOp<mlir::cir::BrOp>(yieldOp, yieldOp.getArgs(),
+                                                   continueBlock);
+    }
+
+    // TODO(cir): stackrestore?
+
+    // Replace the op with values return from the body region.
+    rewriter.replaceOp(scopeOp, continueBlock->getArguments());
+
+    return mlir::success();
+  }
+};
+
+class LoopOpRegionsInlining
+    : public mlir::OpInterfaceRewritePattern<mlir::cir::LoopOpInterface> {
+public:
+  using mlir::OpInterfaceRewritePattern<
+      mlir::cir::LoopOpInterface>::OpInterfaceRewritePattern;
+
+  inline void
+  lowerConditionOp(mlir::cir::ConditionOp op, mlir::Block *body,
+                   mlir::Block *exit,
+                   mlir::PatternRewriter &rewriter) const {
+    mlir::OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPoint(op);
+    rewriter.replaceOpWithNewOp<mlir::cir::BrCondOp>(op, op.getCondition(),
+                                                     body, exit);
+  }
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::LoopOpInterface op,
+                  mlir::PatternRewriter &rewriter) const override {
+    // Setup CFG blocks.
+    auto *entry = rewriter.getInsertionBlock();
+    auto *exit = rewriter.splitBlock(entry, rewriter.getInsertionPoint());
+    auto *cond = &op.getCond().front();
+    auto *body = &op.getBody().front();
+    auto *step = (op.maybeGetStep() ? &op.maybeGetStep()->front() : nullptr);
+
+    // Setup loop entry branch.
+    rewriter.setInsertionPointToEnd(entry);
+    rewriter.create<mlir::cir::BrOp>(op.getLoc(), &op.getEntry().front());
+
+    // Branch from condition region to body or exit.
+    auto conditionOp = cast<mlir::cir::ConditionOp>(cond->getTerminator());
+    lowerConditionOp(conditionOp, body, exit, rewriter);
+
+    // TODO(cir): Remove the walks below. It visits operations unnecessarily,
+    // however, to solve this we would likely need a custom DialecConversion
+    // driver to customize the order that operations are visited.
+
+    // Lower continue statements.
+    mlir::Block *dest = (step ? step : cond);
+    op.walkBodySkippingNestedLoops([&](mlir::Operation *op) {
+      if (isa<mlir::cir::ContinueOp>(op))
+        lowerTerminator(op, dest, rewriter);
+    });
+
+    // Lower break statements.
+    walkRegionSkipping<mlir::cir::LoopOpInterface, mlir::cir::SwitchOp>(
+        op.getBody(), [&](mlir::Operation *op) {
+          if (isa<mlir::cir::BreakOp>(op))
+            lowerTerminator(op, exit, rewriter);
+        });
+
+    // Lower optional body region yield.
```

Comment:
This is the only noticeable change: for some reason `YieldOp` in the loop body was searched only in the first block, which won't work if there are several in there:
```
do {
 ... 
label:
... 
} while(...)

```